llvm/test/tools/sycl-post-link/spec-constants/hierarchy.ll

```diff
@@ -0,0 +1,81 @@
+; For C++ types that come from nested class hierarchy, the LLVM type corresponding
+; to that type seems to match the nested structure. However, it also seems that 
+; when defining a constant for that type, the LLVM value defining the constant has a type 
+; that is different, and is esssentially a flattened out version of the C++ type.
+; For example, this test is IR generated from getting the value of a spec constant
+; of a struct `scary` that has a deep nested hierarchy, but the specialization_id holding
+; the default value of `scary` is a flat struct with all the fields of `scary` flattened out.
+; (compare %struct.scary and @_ZL16scary_spec_const)
+; This test makes that the spec constant pass can handle such cases.
+; (note: IR generated from sycl/test-e2e/SpecConstants/2020/hierarchy.cpp)
+; RUN: sycl-post-link -properties --spec-const=native -S %s -o %t.table
+; RUN: FileCheck %s -input-file=%t_0.ll
+
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-n8:16:32:64-G1"
+target triple = "spir64-unknown-unknown"
+
+%struct.anon = type { i32, i32 }
+%struct.anon.0 = type { i32 }
+%struct.scary = type { %struct.layer4.base, [15 x i8] }
+%struct.layer4.base = type { %struct.layer3.base }
+%struct.layer3.base = type <{ %struct.layer2, [4 x i8], %struct.foo.base }>
+%struct.layer2 = type { %struct.layer1 }
+%struct.layer1 = type { %struct.base }
+%struct.base = type { float, i8, i32, %struct.anon }
+%struct.foo.base = type <{ i32, [4 x i8], [5 x i64], [5 x %struct.anon.0], [5 x i8] }>
+
+@__usid_str = private unnamed_addr constant [44 x i8] c"uid52dfb70f8b72bae7____ZL16scary_spec_const\00", align 1
+@_ZL16scary_spec_const = internal addrspace(1) constant { { float, i8, i32, %struct.anon, [4 x i8], i32, [5 x i64], [5 x %struct.anon.0], [5 x i8], [15 x i8] } } { { float, i8, i32, %struct.anon, [4 x i8], i32, [5 x i64], [5 x %struct.anon.0], [5 x i8], [15 x i8] } { float 0.000000e+00, i8 98, i32 0, %struct.anon zeroinitializer, [4 x i8] undef, i32 3, [5 x i64] [i64 5, i64 0, i64 0, i64 0, i64 0], [5 x %struct.anon.0] [%struct.anon.0 { i32 1 }, %struct.anon.0 { i32 2 }, %struct.anon.0 zeroinitializer, %struct.anon.0 zeroinitializer, %struct.anon.0 zeroinitializer], [5 x i8] c"abc\00\00", [15 x i8] undef } }, align 16
```

Comment:
These `undef` uses are generated by clang, I do not plan on removing them because I want to test that the `SpecConstantsPass` can handle what clang generates.

Reply:
You should not use LLVM IR test format for checking "what clang generates". Clang will generate something different in the future, which this test won't cover. You should use SYCL end-to-end tests to validate end-to-end test behavior.
Please, follow the LLVM community guidelines for the test written in LLVM IR.